crates/bevy_app/src/app.rs

```diff
@@ -70,6 +75,22 @@ struct SubApp {
     runner: Box<dyn Fn(&mut World, &mut App)>,
 }
 
+impl Drop for App {
+    /// Gives a warning if the [`App`] is [`Drop`]ped without being run.
+    fn drop(&mut self) {
+        if !self.has_run {
+            let message = "You have not called `App::run` or `App::update`. Any systems and/or plugins added to this `App` will not run.";
+            dispatcher::get_default(|d| {
+                // If there is no `Subscriber`, `get_default` will return a no-op `Subscriber` called `NoSubscriber`
+                if d.is::<NoSubscriber>() {
+                    eprintln!("WARNING: {}", message);
+                } else {
+                    warn!(warning = message);
+                }
+            });
```

Comment:
there is a `dispatcher::has_been_set()` that is doc hidden but makes this a lot simpler